[1.21.4] Expose vanilla model generators for modded usages

ApexModder opened the pull request:
Updated version of #1696 for `1.21.4`.

While the vanilla providers have been exposed for modded usage, and can be used for more simpler model generation. This is still highly WIP and requires more patches to support our model extensions and generating a model from scratch.

ApexModder committed:
Squash all commits down to 1

Allow modders to generate custom EquipentModels using EquipmentModelProvider
Expose helpers EquipmentModels
Allow modders to use ModelProvider
Expose everything from ItemModelGenerators and BlockModelGenerators
Support modded paths in ModelLocationUtils
Expose more helpers
Support modded paths in ModelTemplate.create
Fix model provider not filtering for mod specific blocks/items
Added test mod to validate ModelProvider and EquipmentModelProvider output
Add note to ModelProvider
Add missing AT for record constructor
Validate model and texture information using ExistingFileHelper
Fix typo in parameter name
Allow using vanilla texture slots in neos model builders
Add custom VariantProperty allowing modders to specify block model builders rather then ResLoc
Allow generating custom model files using vanilla model providers
Allow setting custom properties onto model templates
Allow setting render types for model templates
Fix typo with block state model builder generation
Allow looking up existing model files
Always construct new model template when assigning custom data
Merge overloaded methods into vanilla code
Add more custom data to model templates
Allow setting custom data from TexturedModel
Prepend model folders and append namespaces when missing
Allow specifying render typed by strings (defaulting to "minecraft" namespace)
Add mod and minecraft model path helpers
Fix block models generating empty content
Support setting item transforms on model templates
Fix model being generated into incorrect location
Rename serialization methods
Reorder constructors and parameters
Making them more consistent and the patches easier to read
Remove `VariantProperties.MODEL_BUILDER`
Cleanup `ModelTemplate.createBaseTemplate` patch
Assert existence of existing model files
Allow custom render type names when using String overload
Make `IModelTemplateExtension.withCustomData` publicly visible
Tear out existing file helper
Fix incorrect space indentations
Fix incorrect space indentations *again*
Break all the things
Move texture assignment to TextureSlots
Formatting
Let ModelProvider subclass decide which blocks are relevant
Get compiling on 21.4
Tests model generation still requires updating

ApexModder committed:
Empty commit to trigger workflows

ApexModder commented:
Merged in @XFactHD [fork](https://github.com/ApexModder/NeoForge/pull/1), everything should be exposed and usable. Test mod datagen just needs rewriting to now use vanilla model providers.

ApexModder committed:
Rewrite test mods to now compile and not crash during datagen

ApexModder committed:
Generate patches and apply spotless

ApexModder committed:
Allow generating custom `ClientItem` instances

ApexModder committed:
Move `ITexturedModelExtension` to `client` package

ApexModder committed:
Apply XFacts requested changes

ApexModder committed:
Rename constructor param back to original name

ApexModder committed:
Rename constructor param back to original name *again*

ApexModder committed:
Remove `OverrideBuilder`

Item overrides are defined in the `ClientItem` now

ApexModder committed:
Remove unnecessary semicolon

ApexModder committed:
Fix referencing `BlockModel` when data now moved to `UnbakedModel`

ApexModder commented:
This PR should now be ready for view but important note `ExtendedModelTemplate.Builder` may still get some major reworking to lessen the amount of nesting/`.end` calls.
Thinking of refactoring into lambda style rather than builder style.

Meaning that the following
`ExtendedModelTemplate.builder().element().end()`

would change to something like so
`ExtendedModelTemplate.builder().element(element -> { })`

ApexModder committed:
Convert `ExtendedModelTemplate.Builder` to being lambda based

Split out all inner classes of `ExtendedModelTemplate` into their own classes

ApexModder committed:
Fix nested model issue with `CompositeModelBuilder`

ApexModder committed:
Fix datagen crashes when running game test server

Removes all model gen helpers from `DeferredBlockBuilder` and `DeferredItemBuilder` apart from `withDefaultWhiteModel`

Helpers refactored into proper `addClientProvider` calls within each test

ApexModder committed:
Reapply @IThundxr obj model changes to `NewModelLoaderTest`

ApexModder committed:
Add missing files after running datagen

ApexModder committed:
Reapply PR #1716

XFactHD commented on patches/net/minecraft/client/data/models/ItemModelOutput.java.patch:
```diff
@@ -0,0 +1,10 @@
+ 
+     void copy(Item p_387316_, Item p_387995_);
++
++    default void register(Item item, net.minecraft.client.renderer.item.ClientItem clientItem) {
```
This should have a comment mentioning that Neo pulls this up into the interface to allow mods to control the entire `ClientItem` for an item and not just the `ItemModel.Unbaked` it uses.

marchermans replied:
Please provide JDoc on what this method does or is supposed to do after it now has been pulled up.

We can not expect tools like Parchment to provide this to modders so we should be libraral here with the explanation even if that increases the patch size.

XFactHD commented on patches/net/minecraft/client/data/models/ModelProvider.java.patch:
```diff
@@ -0,0 +1,125 @@
+     private final PackOutput.PathProvider modelPathProvider;
++    public final String modId;
+ 
++    // NeoForge: Use the above constructor passing in their mod id
```
The constructor is below now.

XFactHD commented on patches/net/minecraft/client/data/models/ModelProvider.java.patch:
```diff
@@ -0,0 +1,125 @@
++            this.knownBlocks = knownBlocks;
++        }
++
++        public BlockStateGeneratorCollector() {
```
This constructor should be deprecated.

XFactHD commented on patches/net/minecraft/client/data/models/ModelProvider.java.patch:
```diff
@@ -0,0 +1,125 @@
++            this.knownItems = knownItems;
++        }
++
++        public ItemInfoCollector() {
```
This constructor should be deprecated.

XFactHD commented on patches/net/minecraft/client/data/models/ModelProvider.java.patch:
```diff
@@ -0,0 +1,125 @@
++
++    protected Stream<? extends Holder<Item>> getKnownItems() {
++        return BuiltInRegistries.ITEM.listElements().filter(holder -> holder.getKey().location().getNamespace().equals(modId));
+     }
```
These two methods should have a doc comment mentioning that they should return a stream of all blocks and items respectively for which this generator creates models and maybe also that either can return an empty stream if the generator is used to only generate either block or item models (i.e. if you use a separate provider for pure non-`BlockItem` items).

XFactHD commented on patches/net/minecraft/client/data/models/model/ModelTemplate.java.patch:
```diff
@@ -0,0 +1,31 @@
++        return p_388380_;
++    }
++
++    public JsonObject createBaseTemplate(ResourceLocation p_388380_, Map<TextureSlot, ResourceLocation> map) {
```
This method should have a comment mentioning that it's intended to allow subclasses of `ModelTemplate` to customize serialization.

XFactHD commented on patches/net/minecraft/client/data/models/model/ModelTemplate.java.patch:
```diff
@@ -0,0 +1,31 @@
+         return Streams.concat(this.requiredSlots.stream(), p_387972_.getForced()).collect(ImmutableMap.toImmutableMap(Function.identity(), p_387972_::get));
++    }
++
++    public net.neoforged.neoforge.client.model.generators.template.ExtendedModelTemplateBuilder extend() {
```
I'm not quite happy with the name of this method, it was just the first thing I could think of when I implemented this. I currently can't think of any better name, maybe someone else has a better idea.
This method should ideally also have a short comment explaining what one can do with this.

embeddedt replied:
`createExtendedModelTemplateBuilder` is more descriptive but also ridiculously long.

XFactHD commented on patches/net/minecraft/client/data/models/model/TextureMapping.java.patch:
```diff
@@ -0,0 +1,14 @@
+         return resourcelocation.withPath(p_387396_ -> "item/" + p_387396_ + p_386898_);
+     }
++
++    public TextureMapping copy() {
```
This method should have a comment to indicate that it's added by Neo.

XFactHD commented on src/main/java/net/neoforged/neoforge/client/extensions/IBlockModelGeneratorExtensions.java:
```diff
@@ -0,0 +1,117 @@
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.state.properties.BlockStateProperties;
+
+public interface IBlockModelGeneratorExtensions {
```
All methods in this class should have a short doc comment mentioning that they are standalone versions of the respective methods in `BlockStateGenerators.BlockFamilyProvider`.
The interface's name should also match the target class's name exactly to avoid confusion.

XFactHD commented on src/main/java/net/neoforged/neoforge/client/model/generators/template/ExtendedModelTemplateBuilder.java:
```diff
@@ -0,0 +1,178 @@
+    @Nullable
+    UnbakedModel.GuiLight guiLight = null;
+
+    public static ExtendedModelTemplateBuilder of(ModelTemplate template) {
+        ExtendedModelTemplateBuilder builder = new ExtendedModelTemplateBuilder();
+        builder.parent = template.model;
+        builder.suffix = template.suffix;
+        builder.requiredSlots.addAll(template.requiredSlots);
+        if (template instanceof ExtendedModelTemplate ext) {
+            builder.transforms.copyFrom(ext.transforms);
+            ext.elements.forEach(elem -> builder.elements.add(elem.copy()));
+            builder.customLoader = ext.customLoader != null ? ext.customLoader.copy() : null;
+            builder.rootTransforms.copyFrom(ext.rootTransforms);
+            builder.renderType = ext.renderType;
+            builder.ambientOcclusion = ext.ambientOcclusion;
+            builder.guiLight = ext.guiLight;
+        }
+        return builder;
+    }
+
+    public static ExtendedModelTemplateBuilder builder() {
+        return new ExtendedModelTemplateBuilder();
+    }
+
+    public ExtendedModelTemplateBuilder parent(ResourceLocation parent) {
+        this.parent = Optional.of(parent);
+        return this;
+    }
+
+    public ExtendedModelTemplateBuilder suffix(String suffix) {
+        this.suffix = Optional.of(suffix);
+        return this;
+    }
+
+    public ExtendedModelTemplateBuilder requiredTextureSlot(TextureSlot slot) {
+        this.requiredSlots.add(slot);
+        return this;
+    }
```
These methods need some short doc comments about what they are for, particularly the suffix and texture slot ones as they have a significant impact on the serialization behaviour of the template.

XFactHD commented on src/main/java/net/neoforged/neoforge/client/model/generators/template/ExtendedModelTemplateBuilder.java:
```diff
@@ -0,0 +1,178 @@
+        return this;
+    }
+
+    public ExtendedModelTemplateBuilder transforms(Consumer<TransformsBuilder> action) {
+        action.accept(transforms);
+        return this;
+    }
+
+    public ExtendedModelTemplateBuilder ao(boolean ao) {
+        this.ambientOcclusion = ao;
+        return this;
+    }
+
+    public ExtendedModelTemplateBuilder guiLight(UnbakedModel.GuiLight light) {
+        this.guiLight = light;
+        return this;
+    }
+
+    public ExtendedModelTemplateBuilder element(Consumer<ElementBuilder> action) {
+        Preconditions.checkState(
+                customLoader == null || customLoader.allowInlineElements,
+                "Custom model loader %s does not support inline elements",
+                customLoader != null ? customLoader.loaderId : null);
+        ElementBuilder ret = new ElementBuilder();
+        action.accept(ret);
+        elements.add(ret);
+        return this;
+    }
```
These methods need some short doc comments and the `ao()` method should be renamed to `ambientOcclusion()` for clarity.

XFactHD commented on src/main/java/net/neoforged/neoforge/client/model/generators/template/ElementBuilder.java:
```diff
@@ -0,0 +1,257 @@
+    private int skyLight = 0;
+    private boolean hasAmbientOcclusion = true;
+
+    private static void validateCoordinate(float coord, char name) {
+        Preconditions.checkArgument(!(coord < -16.0F) && !(coord > 32.0F), "Position " + name + " out of range, must be within [-16, 32]. Found: %d", coord);
+    }
+
+    private static void validatePosition(Vector3f pos) {
+        validateCoordinate(pos.x(), 'x');
+        validateCoordinate(pos.y(), 'y');
+        validateCoordinate(pos.z(), 'z');
+    }
```
It would be nice to move these methods to the bottom of the class as they are just noise to someone reading the class to grasp its API surface.

XFactHD commented on src/main/java/net/neoforged/neoforge/client/model/generators/template/ExtendedModelTemplateBuilder.java:
```diff
@@ -0,0 +1,178 @@
+        return this;
+    }
+
+    public ExtendedModelTemplateBuilder rootTransforms(Consumer<RootTransformsBuilder> action) {
+        action.accept(rootTransforms);
+        return this;
+    }
+
+    public ExtendedModelTemplate build() {
+        return new ExtendedModelTemplate(this);
+    }
```
These methods need short doc comments, particularly the root transforms one to differentiate it from the other transforms method.

XFactHD commented on src/main/java/net/neoforged/neoforge/client/model/generators/template/FaceBuilder.java:
```diff
@@ -0,0 +1,126 @@
+    private int skyLight = 0;
+    private boolean hasAmbientOcclusion = true;
+
+    public FaceBuilder cullface(@Nullable Direction dir) {
+        this.cullface = dir;
+        return this;
+    }
+
+    public FaceBuilder tintindex(int index) {
+        this.tintindex = index;
+        return this;
+    }
+
+    /**
+     * Set the texture for the current face.
+     *
+     * @param texture the texture
+     * @return this builder
+     * @throws NullPointerException if {@code texture} is {@code null}
+     */
+    public FaceBuilder texture(TextureSlot texture) {
+        Preconditions.checkNotNull(texture, "Texture must not be null");
+        this.texture = texture;
+        return this;
+    }
+
+    public FaceBuilder uvs(float u1, float v1, float u2, float v2) {
+        this.uvs = new float[] { u1, v1, u2, v2 };
+        return this;
+    }
```
The three methods missing docs need short doc comments.

XFactHD commented on src/main/java/net/neoforged/neoforge/client/model/generators/template/RotationBuilder.java:
```diff
@@ -0,0 +1,70 @@
+        return new BlockElementRotation(origin, axis, angle, rescale);
+    }
+
+    public RotationBuilder copy() {
```
```suggestion
    RotationBuilder copy() {
```

XFactHD commented on src/main/java/net/neoforged/neoforge/client/model/generators/template/TransformsBuilder.java:
```diff
@@ -0,0 +1,42 @@
+        }, LinkedHashMap::new));
+    }
+
+    public void copyFrom(TransformsBuilder builder) {
```
```suggestion
    void copyFrom(TransformsBuilder builder) {
```

XFactHD commented on src/main/java/net/neoforged/neoforge/client/model/generators/template/ElementBuilder.java:
```diff
@@ -0,0 +1,257 @@
+        return this;
+    }
+
+    public ElementBuilder rotation(Consumer<RotationBuilder> action) {
+        if (this.rotation == null) {
+            this.rotation = new RotationBuilder();
+        }
+        action.accept(rotation);
+        return this;
+    }
+
+    public ElementBuilder shade(boolean shade) {
+        this.shade = shade;
+        return this;
+    }
```
These should have short doc comments, particularly shade as that seems to be something that regularly confuses people in relation to AO.

XFactHD commented on src/main/java/net/neoforged/neoforge/client/model/generators/template/RootTransformsBuilder.java:
```diff
@@ -0,0 +1,282 @@
+        return this;
+    }
+
+    public JsonObject toJson() {
```
```suggestion
    JsonObject toJson() {
```

XFactHD commented on src/main/java/net/neoforged/neoforge/client/model/generators/template/RootTransformsBuilder.java:
```diff
@@ -0,0 +1,282 @@
+        return array;
+    }
+
+    public void copyFrom(RootTransformsBuilder other) {
```
```suggestion
    void copyFrom(RootTransformsBuilder other) {
```

XFactHD commented on src/main/java/net/neoforged/neoforge/client/model/generators/template/RotationBuilder.java:
```diff
@@ -0,0 +1,70 @@
+    private float angle;
+    private boolean rescale;
+
+    public RotationBuilder origin(float x, float y, float z) {
```
This method should have a doc comment explaining that this defines the point around which the element will be rotated.

XFactHD commented on src/main/java/net/neoforged/neoforge/client/model/generators/template/RotationBuilder.java:
```diff
@@ -0,0 +1,70 @@
+        return this;
+    }
+
+    public RotationBuilder rescale(boolean rescale) {
```
This method should have a short doc comment explaining what rescaling means in context of the applied rotation.

XFactHD commented on src/main/java/net/neoforged/neoforge/client/model/generators/template/TransformVecBuilder.java:
```diff
@@ -0,0 +1,62 @@
+        // param unused for functional match
+    }
+
+    public TransformVecBuilder rotation(float x, float y, float z) {
+        this.rotation = new Vector3f(x, y, z);
+        return this;
+    }
+
+    public TransformVecBuilder leftRotation(float x, float y, float z) {
+        return rotation(x, y, z);
+    }
+
+    public TransformVecBuilder translation(float x, float y, float z) {
+        this.translation = new Vector3f(x, y, z);
+        return this;
+    }
+
+    public TransformVecBuilder scale(float sc) {
+        return scale(sc, sc, sc);
+    }
+
+    public TransformVecBuilder scale(float x, float y, float z) {
+        this.scale = new Vector3f(x, y, z);
+        return this;
+    }
+
+    public TransformVecBuilder rightRotation(float x, float y, float z) {
+        this.rightRotation = new Vector3f(x, y, z);
+        return this;
+    }
```
These methods should have short doc comments, particularly for `rotation()` and `rightRotation()`.
I think `leftRotation()` should also be removed as it provides no real benefit and `rotation()` being the left rotation can instead be clarified in its docs.

XFactHD reviewed (changes requested):
A couple more comments, primarily about documentation.
When these are adressed, this PR is good to go from my end. The approval and merge has to be done by someone else as this PR adds code I wrote.

ApexModder committed:
Add comment about method being pulled up into iface

ApexModder committed:
Fix issue in comment due to constructor reordering

ApexModder committed:
Deprecate constructor and add note about vanilla/multi-loader compatibility

ApexModder committed:
Add javadocs for `MdoelProvider.getKnown[Block|Item]s`

ApexModder committed:
Add note about reintroducing `ModelTemplate#createBaseTemplate`

ApexModder committed:
Add note about copy method being added by neo

ApexModder committed:
Add javadocs noting why methods were copied from vanilla code

ApexModder committed:
Rename `IBlockmodelGeneratorExtensions` -> `IBlockModelGeneratorsExtension`

ApexModder committed:
Add javadocs describing the vanilla template properties

ApexModder committed:
Inline `TransformsBuilder` into `ExtendedModelTemplateBuilder`

ApexModder committed:
Rename `ao` -> `ambientOcclusion` in `ExtendedModelTemplateBuilder`

ApexModder committed:
Add additional javadocs to `ExtendedMdoelTemplateBuilder`

ApexModder committed:
Move private methods lower down to reduce noise

ApexModder committed:
Add javadocs to methods in `FaceBuilder` without any

ApexModder committed:
Rename `ao` -> `ambientOcclusion` in `FaceBuilder`

ApexModder committed:
Lower visibility of `copy` in `RotationBuilder`

ApexModder committed:
Add javadocs to methods in `ElementBuilder` without any

ApexModder committed:
Rename `ao` -> `ambientOcclusion` in `ElementBuilder`

ApexModder committed:
Lower visibility of `toJson` and `copyFrom` in `RootTransformsBuilder`

ApexModder committed:
Add javadoc to `RotationBuilder#origin`

ApexModder commented:
@XFactHD just pushed changes for all comments aside from `rootTransforms` and `rescale` javadocs as im unsure what these do myself.

Also left the `extend` method as is, as im unsure of a better name myself.

ApexModder committed:
Replace interface patches with interface injection

Matyrobbrt commented on src/main/resources/META-INF/injected-interfaces.json:
```diff
@@ -270,5 +273,14 @@
   "net/minecraft/world/level/biome/BiomeSpecialEffects$GrassColorModifier": [
     "net/neoforged/fml/common/asm/enumextension/IExtensibleEnum"
+  ],
+  "net/minecraft/client/data/models/model/TexturedModel": [
```
Please put all of these in the section above (preferably alphabetically). This section is for enum extension.

ApexModder replied:
Wasnt aware it was split into sections, moved my ifaces to above the first enum extension iface hopefully thats the correct spot.

ApexModder committed:
Reorder `injected-interfaces.json`

ApexModder committed:
Remove `TransformVecBuilder#leftRotation`

ApexModder committed:
Add javadocs to `TransformVecBuilder`

ApexModder committed:
Add additional javadocs to `ExtendedModelTemplateBuilder`

XFactHD commented on patches/net/minecraft/client/data/models/ModelProvider.java.patch:
```diff
@@ -0,0 +1,163 @@
+-            modelprovider$iteminfocollector.save(p_387857_, this.itemInfoPathProvider)
++                modelprovider$blockstategeneratorcollector.save(p_387857_, this.blockStatePathProvider),
++                modelprovider$simplemodelcollector.save(p_387857_, this.modelPathProvider),
++                modelprovider$iteminfocollector.save(p_387857_, this.itemInfoPathProvider)
```
Please revert this, it's purely a formatting change.

XFactHD commented on src/main/java/net/neoforged/neoforge/client/model/generators/template/TransformVecBuilder.java:
```diff
@@ -0,0 +1,79 @@
+    /**
+     * Sets the rotation for this transformation.
+     * <p>
+     * This is rotation is considered the "left rotation" when also using {@linkplain #rightRotation(float, float, float)}.
```
```suggestion
     * This rotation is considered the "left rotation" when also using {@linkplain #rightRotation(float, float, float)}.
```

XFactHD commented on patches/net/minecraft/client/data/models/ItemModelOutput.java.patch:
```diff
@@ -0,0 +1,11 @@
+ 
+     void copy(Item p_387316_, Item p_387995_);
++
++    // Neo: Pulled up from 'ModelProvider.ItemInfoCollector' to give modders full control over the 'ClientItem' over the 'ItemModel.Unbacked' in '#accept'
```
```suggestion
+    // Neo: Pulled up from 'ModelProvider.ItemInfoCollector' to give modders full control over the 'ClientItem' instead of just the 'ItemModel.Unbacked' in '#accept'
```

ApexModder committed:
Fix silly formatting mistake in `ModelProvider` patch

ApexModder committed:
Remove extra `is` from javadoc

ApexModder committed:
Update comment in `ItemModelOutput`

ApexModder committed:
Add javadoc to `RotationBuilder#rescale`

ApexModder committed:
Fix more formatting issues in `ModelProvider` patch

ApexModder committed:
Fix typos in javadoc

ApexModder committed:
Reword javadoc

ApexModder committed:
Add comment to `ModelTemplate#extend` describing what modders can do with it

ApexModder commented:
PR should be ready for review and merge, only unresolved comment should be xfacts one about `ModelTemplate#extend` but we are both unsure of a more fitting name.

marchermans commented on src/main/java/net/neoforged/neoforge/client/extensions/IBlockModelGeneratorsExtension.java:
```diff
@@ -0,0 +1,151 @@
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.state.properties.BlockStateProperties;
+
+public interface IBlockModelGeneratorsExtension {
```
This extension is not strictly necessary.
It is possible to create a custom BlockFamily that is incomplete and use it to perform these generations

marchermans replied:
Please remove it.

marchermans commented on src/main/java/net/neoforged/neoforge/common/data/ExistingFileHelper.java:
When this originally was discussed we realised that this class is a very large hack that does not work well anymore in modern data gen.

We ripped it out of model and blockstate generation now, and should remove it from the rest as well.

Please mark this as Deprecated without removal date, so we keep an eye on this.

ChampionAsh5357 commented on patches/net/minecraft/client/data/models/ItemModelOutput.java.patch:
```diff
@@ -0,0 +1,11 @@
+ 
+     void copy(Item p_387316_, Item p_387995_);
++
++    // Neo: Pulled up from 'ModelProvider.ItemInfoCollector' to give modders full control over the 'ClientItem' instead of just the 'ItemModel.Unbacked' in '#accept'
```
```suggestion
+    // Neo: Pulled up from 'ModelProvider.ItemInfoCollector' to give modders full control over the 'ClientItem' instead of just the 'ItemModel.Unbaked' in '#accept'
```

ChampionAsh5357 commented on src/main/java/net/neoforged/neoforge/client/extensions/ITexturedModelExtension.java:
```diff
@@ -0,0 +1,30 @@
+import net.minecraft.client.data.models.model.TexturedModel;
+
+public interface ITexturedModelExtension {
+    default TexturedModel updateTemplate(UnaryOperator<ModelTemplate> modifier) {
+        return new TexturedModel(self().getMapping(), modifier.apply(self().getTemplate()));
+    }
```
I feel like this needs some documentation to explain why `updateTemplate` makes sense, since the additional data used by the custom loaders is set on the template itself rather than the textures slots.

ChampionAsh5357 commented on src/main/java/net/neoforged/neoforge/client/model/generators/loaders/DynamicFluidContainerModelBuilder.java:
I feel like this could be moved to an extension method on `ItemModelGenerators` rather than being outright removed as it still is a neoforge-added item model.

XFactHD replied:
The only reason custom model loaders have builders is because they aren't reversible and therefore not serializable. The dynamic fluid container model is a custom `ItemModel` now though and the associated `ItemModel.Unbaked` can just be instantiated and passed to `ItemModelOutput#register()`.

ChampionAsh5357 replied:
I understand. I just figured it would probably be easier to have a convenience method to create them, similar to all the methods in the model generator.

ChampionAsh5357 commented on src/main/java/net/neoforged/neoforge/client/model/generators/template/CustomLoaderBuilder.java:
```diff
@@ -76,4 +81,8 @@ public JsonObject toJson(JsonObject json) {
         return json;
     }
+
+    protected static void serializeNestedTemplate(ModelTemplate template, TextureMapping textures, Consumer<JsonElement> consumer) {
+        template.create(DUMMY, textures, (id, jsonSup) -> consumer.accept(jsonSup.get()));
+    }
```
Documentation? I don't know what this would be for.

ApexModder replied:
Method was unused and thus has been removed, was told its better practice to now reference external files rather than inlining models like we used to, which this was used for in the past.

ChampionAsh5357 replied:
Is this supposed to be removed? Basing this off your previous comment.

XFactHD replied:
Apex did remove it at first but it turned out that we'll need it for a subsequent PR, so it was added back and documented.

ChampionAsh5357 reviewed (changes requested):
A few minor things excluding what Orion said. I also removed my comments about the `ExistingFileHelper` if we are planning to strip it out.

ApexModder committed:
Convert comment to jdoc

ApexModder committed:
Tear out `BlockFamily` wrapper extension methods

ApexModder committed:
Remove `IBlockModelGeneratorsExtension`

ApexModder committed:
Mark `ExistingFileHelper` for removal

ApexModder committed:
Add jdocs to `ITexturedModelExtension#updateTemplate`

ApexModder committed:
Nuke unused method and field

ApexModder committed:
Add method to allow generating family models for existing blocks

neoforged-automation[bot] committed:
Merge branch '1.21.x' into pr/vanilla-datagen-21.4

ApexModder committed:
Revert "Nuke unused method and field"

This reverts commit f9eb871c80ec65771ed18c6eec03382cd0fe7068.

ApexModder committed:
Add jdoc to `CustomLoaderBuilder#serializeNestedTemplate`

neoforged-automation[bot] committed:
Merge branch '1.21.x' into pr/vanilla-datagen-21.4

Soaryn commented on src/main/java/net/neoforged/neoforge/client/model/generators/template/FaceBuilder.java:
```diff
@@ -0,0 +1,135 @@
+        return new BlockElementFace(cullface, tintindex, texture.toString(), new BlockFaceUV(uvs, rotation.rotation), new ExtraFaceData(this.color, this.blockLight, this.skyLight, this.hasAmbientOcclusion), new MutableObject<>());
+    }
+
+    FaceBuilder copy() {
```
Is this missing the textureslot copy? Seems like we'd need to make a new template with the textures manually otherwise

Soaryn commented on src/main/java/net/neoforged/neoforge/client/model/generators/template/FaceBuilder.java:
```diff
@@ -0,0 +1,135 @@
+    @Nullable
+    private Direction cullface;
+    private int tintindex = -1;
+    @Nullable
```
Can it really be classified as `Nullable` if it errors when it is null? 
`build()` calls `.toString()`

ApexModder replied:
I mean yeah, internally it is null cause it has no initial value (what texture slot would you default every face to?) but all attempts for users to set the texture are null checked and the final `build` call also null checks the texture, making it a required non null field.

this is also how face builder has always functioned in the legacy modelgen code iirc and no ones had issues with it before as far as im aware.

ApexModder committed:
Copy texture slot when invoking `FaceBuilder#copy`

ApexModder committed:
Copy color when invoking `FaceBuilder#copy`

ApexModder committed:
Make `CustomLoaderBuilder#copy` package-private

Add jdoc to `CustomLoaderBuilder#copy` stating that it must a deep-copy of the current builder instance

ChampionAsh5357 commented on patches/net/minecraft/client/data/models/BlockModelGenerators.java.patch:
```diff
@@ -0,0 +1,19 @@
+         }
+     }
++
++    /**
++     * Neo: create a {@link BlockModelGenerators.BlockFamilyProvider} which re-uses the existing model of the given full
++     * block instead of creating a model and blockstate file for it. Intended for use cases where the full block is
++     * separately generated or otherwise exists such as when a dummy {@link BlockFamily} is used to create additional
++     * variants for existing vanilla block families
++     */
++    public BlockModelGenerators.BlockFamilyProvider familyWithExistingFullBlock(Block fullBlock) {
++        var provider = new BlockModelGenerators.BlockFamilyProvider(TextureMapping.cube(fullBlock));
++        provider.fullBlock = ModelLocationUtils.getModelLocation(fullBlock);
++        return provider;
++    }
+ }
```
Is there a reason this has to be a patch rather than a extension method? Or are we not planning on adding any common usages to the base generators?

XFactHD replied:
It would (at least for now) be the only extension and doing it as a patch is simpler as an extension method would require exposing the `BlockFamilyProvider#fullBlock` field.

ChampionAsh5357 commented on patches/net/minecraft/client/data/models/BlockModelGenerators.java.patch:
```diff
@@ -0,0 +1,19 @@
+         }
+     }
++
++    /**
++     * Neo: create a {@link BlockModelGenerators.BlockFamilyProvider} which re-uses the existing model of the given full
++     * block instead of creating a model and blockstate file for it. Intended for use cases where the full block is
++     * separately generated or otherwise exists such as when a dummy {@link BlockFamily} is used to create additional
++     * variants for existing vanilla block families
++     */
++    public BlockModelGenerators.BlockFamilyProvider familyWithExistingFullBlock(Block fullBlock) {
++        var provider = new BlockModelGenerators.BlockFamilyProvider(TextureMapping.cube(fullBlock));
++        provider.fullBlock = ModelLocationUtils.getModelLocation(fullBlock);
++        return provider;
++    }
+ }
```
Is there a reason this has to be a patch rather than a extension method? Or are we not planning on adding any common usages to the base generators?

ChampionAsh5357 reviewed (changes requested):
Just some quick comments based on my previous ones.

sciwhiz12 commented on src/main/java/net/neoforged/neoforge/common/data/ExistingFileHelper.java:
```diff
@@ -46,6 +45,7 @@
  * extra resource packs passed in via the {@code --existing} argument,
  * or mod resources via the {@code --existing-mod} argument.
  */
+@Deprecated(forRemoval = true, since = "1.21.4")
```
Why is this deprecated for removal? There should likely be a comment explaining why, and what alternatives (if any) should be used instead.

ApexModder replied:
Existing file helper afaik is officially dead and marked for removals with no planned alternative to be implemented. There have been talks however about a follow up pr to go over to fully make EFH optional/nullable everywhere.

sciwhiz12 commented on src/main/resources/META-INF/injected-interfaces.json:
```diff
@@ -271,4 +279,4 @@
   "net/minecraft/world/level/biome/BiomeSpecialEffects$GrassColorModifier": [
     "net/neoforged/fml/common/asm/enumextension/IExtensibleEnum"
   ]
-}
+}
```
Is this removal of the newline intentional?

sciwhiz12 commented on patches/net/minecraft/client/data/models/ModelProvider.java.patch:
```diff
@@ -0,0 +1,140 @@
+     private final PackOutput.PathProvider modelPathProvider;
++    public final String modId;
+ 
++    // NeoForge: Use the below constructor passing in your mod id
```
Convention is for comments in patches to use `Neo:`. Also, reword the comment to avoid using position, and instead refer to the additional parameter.
```suggestion
+    // Neo: Use the constructor which accepts a mod ID.
```

sciwhiz12 commented on patches/net/minecraft/client/data/models/ModelProvider.java.patch:
```diff
@@ -0,0 +1,140 @@
++            this.knownBlocks = knownBlocks;
++        }
++
++        @Deprecated // Neo: Provided for vanilla/multi-loader compatibility, use above constructor
```
Reword the comment to avoid using position, and instead refer to the additional parameter.
```suggestion
+        @Deprecated // Neo: Provided for vanilla/multi-loader compatibility. Use constructor with Supplier parameter.
```

sciwhiz12 commented on patches/net/minecraft/client/data/models/ModelProvider.java.patch:
```diff
@@ -0,0 +1,140 @@
++            this.knownItems = knownItems;
++        }
++
++        @Deprecated // Neo: Provided for vanilla/multi-loader compatibility, use above constructor
```
Reword the comment to avoid using position, and instead refer to the additional parameter.
```suggestion
+        @Deprecated // Neo: Provided for vanilla/multi-loader compatibility. Use constructor with Supplier parameter.
```

sciwhiz12 commented on patches/net/minecraft/client/data/models/model/ModelLocationUtils.java.patch:
```diff
@@ -0,0 +1,18 @@
+     @Deprecated
+     public static ResourceLocation decorateBlockModelLocation(String p_388520_) {
+-        return ResourceLocation.withDefaultNamespace("block/" + p_388520_);
++        // NeoForge: Use ResourceLocation.parse to support modded paths
```
Convention is for comments in patches to use `Neo:`.
```suggestion
+        // Neo: Use ResourceLocation.parse to support modded paths
```

sciwhiz12 commented on patches/net/minecraft/client/data/models/model/ModelLocationUtils.java.patch:
```diff
@@ -0,0 +1,18 @@
+ 
+     public static ResourceLocation decorateItemModelLocation(String p_387226_) {
+-        return ResourceLocation.withDefaultNamespace("item/" + p_387226_);
++        // NeoForge: Use ResourceLocation.parse to support modded paths
```
Convention is for comments in patches to use `Neo:`.
```suggestion
+        // Neo: Use ResourceLocation.parse to support modded paths
```

sciwhiz12 commented on tests/src/main/java/net/neoforged/neoforge/debug/VanillaDataGenTest.java:
```diff
@@ -0,0 +1,68 @@
+
+@ForEachTest(groups = "vanilla_data_gen", side = Dist.CLIENT)
+public interface VanillaDataGenTest {
+    @TestHolder(description = "Tests the the patched vanilla model generators work for modded usages")
```
Typo: double "the".
```suggestion
    @TestHolder(description = "Tests the patched vanilla model generators work for modded usages")
```

ApexModder committed:
Fix double `the` in annotation parameter

ApexModder committed:
Fix up issues in comments

ApexModder committed:
Add deprecation jdoc to `ExistingFileHelper`

neoforged-automation[bot] committed:
Merge branch '1.21.x' into pr/vanilla-datagen-21.4

neoforged-automation[bot] committed:
Merge branch '1.21.x' into pr/vanilla-datagen-21.4
